File: src/java/crc/pia/Transaction.java
// Transaction.java
// $Id$
// (c) COPYRIGHT Ricoh California Research Center, 1997.


/**
 * Transactions generalize the HTTP classes Request and Response.
 * They are used in the rule-based resolver, which associates 
 * transactions with the interested agents.
 *
 * A Transaction has a queue of ``handlers'', which are called
 * (from the $transaction->satisfy() method) after all agents
 * have acted on it.  At least one must return true, otherwise
 * the transaction will ``satisfy'' itself.
 *
 * In a proper implementation, Transaction would be a subclass of 
 * DS::Thing, and Response and Request would be subclasses of it.
 * It's done backwards here in order to re-use existing libraries.
 *
 */

package crc.pia;
import java.util.Enumeration;
import java.net.URL;
import java.io.IOException;

import crc.ds.Thing;
import crc.ds.Queue;
import crc.pia.Machine;
import crc.pia.Content;
import crc.tf.Registry;

public class Transaction extends Thing{
  /**
   * Attribute index - if true transaction is a response.
   *
   */
  public boolean isResponse = false ;

  /**
   * Attribute index - if true transaction is a request.
   *
   */
  public boolean isRequest = false;


  /**
   * Attribute index - from machine -- the machine that originates the request.
   *
   */
  protected Machine fromMachine;

  /**
   * Attribute index - to machine -- target machine that will process the request.
   *
   */
  protected Machine toMachine;

  /**
   * Attribute index - first line -- can be request or response line
   */
  protected String firstLine;

  /**
   * Attribute index - content obj of this transaction.
   *
   */
  protected Content contentObj;

  /**
   * Attribute index - header obj of this transaction.
   */
  protected Headers headerObj;

  /**
   * Attribute index - queue of handlers.
   *
   */
  protected Queue handlers;

  /**
   * Attribute index - controls to be insert into a response
   *
   */
  protected Vector controls = new Vector();


  /**
   * Attribute index - store request URL string
   */
  protected Transaction requestTran;

  /**
   * Below are interfaces related to Request transaction.  
   *
   */


  /**
   *  @returns the content length for this request, or -1 if not known. 
   */
  public int contentLength(){
    int res = -1;
    Content c = contentObj();
    if( c && c.headers() != null )
      res = c.headers().contentLength();
    return res;
  } 

  /**
   *  @returns the content type for this request, or null if not known. 
   *
   */
  public String contentType(){
    String res = null;
    Content c = contentObj();
    if( c && c.headers() != null )
      res =  c.headers().contentType();
    return res;
  }

  /**
   * @returns the value of a header field, or null if not known. 
   *
   */
  public String header(String name){
    String res = null;
    Content c = contentObj();
    if( c && c.headers() != null )
      res = c.headers().header( name );
    return res;
  }

  /**
   * @returns all header information as string. Machine.java uses this.
   *
   */
   public String headerAsString(){
     String res = null;
     Content c = contentObj();
     if( c && c.headers() != null )
       res = c.headers().toString();
     return res;
   }

  /**
   * @returns the request method. 
   */
  public String method() {
    if(isResponse())
      return null;

    String m = null;
    Object result =  is("Method");
    if( result )
      m = (String) result;

    return m;   
  }

  /**
   *   @returns the protocol of the request. 
   *
   */
  public String protocol() {
    return "HTTP";
  }
  
  /**
   *   @returns the host name of the remote agent, or null if not known. 
   *
   */
  public String remoteHost() {
    String res = null;

    Content c = contentObj();
    if( c && c.headers() != null )
      res = c.headers().getValue( "Host" );
    return res;
  }
  
  /**
   *   @returns the full request URI. 
   *
   */    
  public URL requestURL(){
    if(isResponse())
      return null;

    URL myUrl = null;
    Object result = is("Url");
    if( result )
      myUrl = (URL) result;

    return myUrl;   
  }

  /**
   * Set header info.
   *
   */
  public void setHeader(String key, String value){
    Content c = contentObj();
    if( c && c.headers() != null )
      res = c.headers().setValue(key, value);
   }


  /**
   * Interfaces related to Response.
   */


  /**
   *   @returns the status code for this response. 
   *
   */
  public int statusCode(){
    int s = -1;
    if(isRequest())
      return s;

    Object result =  is("Statuscode");
    if( result )
      s = (int) result;

    return s;
  }
  
  /**
   *   Sets the content length for this response. 
   *
   */
  public void setContentLength(int len){
    Content c = contentObj();
    if( c && c.headers() != null )
      res = c.headers().setContentLength( len );
  }
  
  /**
   *   Sets the content type for this response. 
   *
   */
  public void setContentType(String type){
    Content c = contentObj();
    if( c && c.headers() != null )
      res = c.headers().setContentType( type ); 
  }

  // code from jigsaw
  /**
   * Get the standard HTTP reason phrase for the given status code.
   * @param status The given status code.
   * @return A String giving the standard reason phrase, or
   * <strong>null</strong> if the status doesn't match any knowned error.
   */

    public String standardReason(int status) {
	int category = status / 100;
	int catcode  = status % 100;
	switch(category) {
	  case 1:
	      if ((catcode >= 0) && (catcode < msg_100.length))
		  return HTTP.msg_100[catcode];
	      break;
	  case 2:
	      if ((catcode >= 0) && (catcode < msg_200.length))
		  return HTTP.msg_200[catcode];
	      break;
	  case 3:
	      if ((catcode >= 0) && (catcode < msg_300.length))
		  return HTTP.msg_300[catcode];
	      break;
	  case 4:
	      if ((catcode >= 0) && (catcode < msg_400.length))
		  return HTTP.msg_400[catcode];
	      break;
	  case 5:
	      if ((catcode >= 0) && (catcode < msg_500.length))
		  return HTTP.msg_500[catcode];
	      break;
	}
	return null;
    }


    /**
     * Set this reply status code.
     * This will also set the reply reason, to the default HTTP/1.1 reason
     * phrase.
     * @param status The status code for this reply.
     */

    public void setStatus(int status) {
	if ((statusCode() != this.status) || (reason() == null))
	    assert("Reason", standardReason(status) );
	assert("Statuscode", status );
    }

    /**
     * Get the reason phrase for this reply.
     * @return A String encoded reason phrase.
     */

    public String reason() {
      if( isRequest() )
	return null;

      String s = null;
      Object result =  is("Reason");
      if( result )
	s = (String) result;

      return result;
    }

    /**
     * Set the reason phrase of this reply.
     * @param reason The reason phrase for this reply.
     */

    public void setReason(String reason) {
	assert("Reason", reson);
    }


  /**
   * @return requested transaction
   */
  public Transaction requestTran(){
    if( requestTran )
      return requestTran;
    else
      return null;
  }

  /**
   * true if this transaction is a response
   */
  public boolean isResponse(){
    return isResponse;
  }

  /**
   * true if this transaction is a request
   */
  public boolean isRequest(){
    return isRequest;
  }


  /**
   * queue -- returns handler list
   * 
   */ 
  protected Enumeration queue(){
    return handlers.queue;
  }
 
  /**
   * shift -- remove and return the handler at front of list .
   * If there is no transaction returns null.
   */
  public Object shift(){
    return handlers.shift();
  }

  /**
   * unshift -- put a handler to the front of the list. 
   * @return the number of handlers
   */ 
  public int unshift( Object obj ){
    return handlers.unshift( obj );
  }

  /**
   * push -- push a handler onto the end of the list. 
   * @return the number of elements
   */  
  public int push( Object obj ){
    return handlers.push( obj );
  }

  /**
   * pop -- removes a handler from the back of the queue and returns it. 
   * @return the number of elements
   */ 
  public Object pop(){
    return handlers.pop();
  }

  /**
   * Number of handlers in queue
   *
   */
  public int size(){
    return handlers.size();
  }

  /**
   * return a ds's Feature object 
   */

  public Features features (){
    // return from Thing's _features
    return features;
  }

  /**
   * set Features object
   */
  public void setFeatures(Features features){
    if( features )
      this.features = features;
  }

  /**
   * Get the value of the named feature.  If does not exist,
   * compute it and return the value
   */
  public Object is( String name ) {
    return features.test(name, this);
  }

  /**
   * Test a named feature and return a boolean.
   */
  public boolean test( String name ) {
    return features.test(name, this);
  }

  /**
   * Compute and assert the value of the given feature.
   * Can be used to recompute features after changes
   */
  public Object compute( String name ){
    return features.compute(name, this);
  }

  /**
   * assert a given feature with the given value
   */
  public void assert( String name, Object value ) {
    features.assert($feature, $value);
  }

  /**
   * deny a given feature
   */
  public void deny( String name ) {
    features.deny(name);
  }

  public boolean has( String name ) {
    return features.has(name);
  }

  public Object computeFeature( String featureName ) throws UnknownNameException{
    UnaryFunctor c;
    try{
     c = (UnaryFunctor)Registry.calculatorFor( featureName );
     return c.execute();
    }catch(UnknownNameException e){
      // log here
    }
    
  }

  /**
   * Accessing function to content object
   * 
   */
  public Content contentObj(){
    return contentObj;
  }


  /**
   * temporary treatment of content objects
   * while transitioning to full objectIvity...
   */
  protected void initializeContent( Content ct ){
    //content source set in fromMachine method
    contentObj = ct;
  }
  /**
   * parse the request line to get method, url, http's major and minor version numbers
   */
  protected void parseRequestLine(){
    if( !isRequest() ) return;
    StringTokenizer tokens = new StringTokenizer(firstLine, " ");
    String zmethod = tokens.nextToken();
    
  }

  /**
   * temporary treatment of content objects
   * while transitioning to full objectIvity...
   * Append previous transaction's content to this trasaction content
   */
  protected void initialize(HeaderFactory hf, ContentFactory cf ){
    //content source set in fromMachine method
    InputStream in;

    try{
      in = fromMachine().inputStream();

      DataInputStream input = new DataInputStream( in );
      firstLine = input.readLine();

      headerObj  = hf.createHeader( in );
      parseRequestLine();
      parseResponseLine();
      contentObj = cf.createContent( in );
    }catch(IOException e){
    }
  }


  /**
   * fromMachine returns machine that initialize the request.
   * 
   */
  public Machine fromMachine(){
    return fromMachine;
  }
  
  /**
   * fromMachine sets to fromMachine the machine that initializeed the request.
   * 
   */
  public void fromMachine(Machine machine){
    if( machine ){
      fromMachine = machine;
      contentObj.source( machine.inputStream() );
      machine.sLength( contentObj.contentLength() );
    }
  }

  /**
   * toMachine get toMachine-- the machine that is the target of the request.
   * 
   */
  public Machine toMachine(){
    String host;
    URL url = requestURL();
    String urlString;

    if(!toMachine && is_request){
      urlString = url.getFile();
      if( urlString )
	host = url.getHost();
      if( host )
	toMachine = new Machine( host );
    }
    return toMachine;
  }

  /**
   * toMachine sets toMachine-- the machine that is the target of the request.
   * 
   */
  public void toMachine(Machine machine){
    if( machine ){
      toMachine = machine;
    }
  }



  /**
   * handle -- A transaction can handle a request by pushing itself
   * onto the given resolver.  This allows agents to push
   * responses onto a *transaction* to be handled.  We return
   * success, indicating that the request has been satisfied.
   *
   */
  public boolean handle( Resolver resolver ){
    // deliver responses before processing more request
    resolver.unshift( this );
    return true;
  }

  /**
   * sendResponse -- Utilities to actually respond to a transaction, get a request, 
   * or generate (return) an error response transaction.
   *
   * These pass the resolver down to the Machine that actually does the 
   * work, because it might belong to an agent.
   *
   */
  public void sendResponse( Resolver resolver ){
    /*
     * Default handling for a response:
     * send it to the toMachine.  If the destination is not a reference 
     * to a machine, the response just gets dropped.  'Nowhere' is a good
     * non-reference to use in this case.
     */
    Machine machine = toMachine();
    
    if ( machine ){
      machine.sendResponse(this, resolver);
    }
    else{
      System.out.println( "dropping  response to $machine\n" );
      return;
    }
    
  }

  /**
   * handleRequest -- Default handling for a request:
   * ask the destination machine to get it.
   * complain if there's no destination to ask.
   */
  public Transaction handleRequest( Resolver resolver ){
    Machine destination = toMachine();
    Transaction response;


    if(!destination)
      return errorResponse();
    response = destination.getRequest( this, resolver );
    return response;
  }
  
  /**
   * errorResponse -- Return a "not found" error for a request with no destination.
   *
   *
   */
  protected Transaction errorResponse(){
    Transaction response = new Transaction( pia.thisMachine, fromMachine() );
    
    response.setStatus( 404, "Not found" );
    response.setContentType( "text/plain" );
    response.setContent("Agency could not find " + requestURL() + ".\n");
    return response;
  }


  /**
   * Satisfying transactions:
   * A transaction can handle a request by pushing itself
   * onto the given resolver.  This allows agents to push
   * responses onto a *transaction* to be handled.  We return
   * success, indicating that the request has been satisfied.
   *
   */
  public void satisfy(Resolver resolver) {
    Object obj;
    boolean satisfied = false;

    Enumeration e = handlers.queue();
    while( e.hasMoreElements() ){
      obj = e.nextElement();
      if( obj instanceof Thing  ){
	Thing agentortran = (Thing)obj;
	if ( agentortran.handle(this, resolver) == true )
	  satisfied = true;
      }
      else{
	if ( obj instanceof boolean  ){
	  boolean result = (boolean) obj;
	  if( result )
	    satisfied = true;
	}
      }
    }

    //If still not satisfied, do the default thing:
    //  send a response or get a request.
    if(!satisfied){
      if( isResponse() )
	sendResponse( resolver );
      else {
	if ( isRequest() )
	  resolver.push( handleRequest( resolver ) );
      }
    }

  }

  /**
   * controls -- Add controls (buttons,icons,etc.) for agents to this response
   * actual final form determined by machine
   * NOTE: not implemented
   */
  public void addControl( Thing aThing ){
    controls.addElement( aThing );
  }

  /**
   * controls -- Add controls (buttons,icons,etc.) for agents to this response
   * actual final form determined by machine
   * NOTE: not implemented
   */
  public Thing[] controls(){
    int size = 0;

    size = controls.size();
    if( size > 0 ){
      Thing[] c = new Thing[ size ];
      for(int i = 0; i < size; i++ ) 
	c[i] = (Thing) controls[i];
      return c;
    }
    return null;
  }

  /**
   * A request transaction 
   * 
   */
  public Transaction( Machine from ){
    handlers = new Queue();
    HeaderFactory hf  = new HeaderFactory();
    ContentFactory cf = new ContentFactory();
    
    initialize( hf, cf );
    isRequest = true ;
    fromMachine( from );
    toMachine( null );
  }

  /**
   * A Request transaction
   */
  public Transaction( Machine from, Content ct ){
    handlers = new Queue();

    initializeContent( ct );
    isRequest =  true;
    fromMachine( from );
    toMachine( null );

  }

  /**
   * A Response transaction
   */
  public Transaction(Transaction t, Content ct ){
    handlers = new Queue();

    requestTran = t;
    initializeContent( ct );
    isResponse = true;
    fromMachine( t.toMachine() );
    toMachine( t.fromMachine() );
  }
} 

File: src/perl/transaction.pl
###### class TRANSACTION
###	$Id$
###
###	Transactions generalize the HTTP classes Request and Response.
###	They are used in the rule-based resolver, which associates 
###	transactions with the interested agents.
###
###	A transaction has a queue of ``handlers'', which are called
###	(from the $transaction->satisfy() method) after all agents
###	have acted on it.  At least one must return true, otherwise
###	the transaction will ``satisfy'' itself.

package TRANSACTION;

push(@ISA,HTTP::Request);
push(@ISA,HTTP::Response);
push(@ISA,HTTP::Message);


###take request or response and make transaction
sub new {
    my($class,$request,$from,$to)=@_;
    my $old=ref($request);
    bless $request,$class;
    $request->is_request(1) if $old eq 'HTTP::Request';
    $request->is_response(1) if $old eq 'HTTP::Response';

    $$self{queue} = [];

    print "Making transaction from $old\n" if $main::debugging;

    $request->from_machine($from);
    $request->to_machine($to);
    my $type=$request->method;
    print "  type is $type\n"  if $main::debugging;    
    if(($type eq "POST" || $type eq "PUT") && $request->is_request()){
	$request->read_content($type);
	$request->compute_form_parameters() if $type eq 'POST';
    }
    return $request;
}



sub from_machine{
    my($self,$machine)=@_;
    $$self{_from_machine}=$machine if defined $machine;
    return $$self{_from_machine};
}

sub to_machine{
    my($self,$machine)=@_;
    $$self{_to_machine}=$machine if defined $machine;
    return $$self{_to_machine};
}

sub is_response{
    my($self,$argument)=@_;
    $$self{_response_Boolean}=$argument if defined $argument;
    return $$self{_response_Boolean};
}

sub is_request{
    my($self,$argument)=@_;
    $$self{_request_Boolean}=$argument if defined $argument;
    return $$self{_request_Boolean};
}

### Queue handlers.

sub queue{
    my($self)=@_;
    $$self{queue} = [] if !defined $$self{queue};
    return $$self{queue};
}

sub shift{
    my($self)=@_;
    return shift(@{$self->queue()});
}
sub unshift{
    my($self,$argument)=@_;
    return unshift(@{$self->queue()},$argument);
}
sub push{
    my($self,$argument)=@_;
    return push(@{$self->queue()},$argument);
}
sub pop{
    my($self)=@_;
    return pop(@{$self->queue()});
}

###### $transaction->handle($request, $resolver)
###
###	A transaction can handle a request by pushing itself
###	onto the given resolver.  This allows agents to push
###	responses onto a *transaction* to be handled.  We return
###	success, indicating that the request has been satisfied.
###
sub handle {
    my ($self, $request, $resolver) = @_;

    $resolver -> push($self);
    return 1;
}


###### $transaction->satisfy($resolver)
###
###	Calls $agent->handle($self, $resolver) for every agent in its
###	queue.  (If the agent is not a reference, it is taken as a
###	boolean result; this allows an ``act-on'' agent to mark the
###	transaction as satisfied by pushing a 1 onto it.
###
sub satisfy {
    my ($self, $resolver) = @_;
    my $satisfied = 0;

    foreach $agent (@{$self->queue()}) {
	if (ref($agent)) {
	    $satisfied = 1 if $agent->handle($self, $resolver);
	} else {
	    $satisfied = 1 if $agent;
	}
    }

    if (! $satisfied) {
	if ($self->is_response()) {
	    $self->send_response();
	} elsif ($self->is_request()) {
	    print "pushing error_response\n" if $main::debugging;
	    $resolver->push($self->error_response());
	}
    }
}


############################################################################
###
### Utilities to actually respond to a transaction, or generate
###	(return) an error response transaction.
###

sub send_response {
    my($reply)=@_;
    my $machine=$reply->to_machine();

    if (ref($machine)) {
	print "sending response.\n" if $main::debugging;
	my $status=$machine->send_response($reply) if ref($machine);
	return $status;
    } else {
	print "sending response to $machine\n" if $main::debugging;
	return;
    }
}

sub error_response{
    my($self)=@_;
    my $response=HTTP::Response->new(&HTTP::Status::RC_NOT_FOUND, "not found");
    $response->content_type("text/plain");    
    my $url=$self->url()->as_string();
    
    $response->content("Agency could not find $url\n");

    $response=TRANSACTION->new($response,
			       $main::this_machine,
			       $argument->from_machine());
    return $response;
}


###### transaction->read_content($type)
###
###	Read the content from the from_machine's stream.
###
###	=== This needs work: ===
###	 o  Must make sure it works with both requests and responses
###	 o  Must make sure it works OK if repeated! (idempotent)
###	 o  Must handle the case where the content is in a file.
###	 o  Should be able to fetch only the <HEAD> of an html file
###	 o  Must work if content_length was not provided.
###
sub read_content{
    my($self,$type)=@_;
    
    my $bytes=$self->content_length;
        
    my $content;
    my $from=$self->from_machine();
    return unless defined $from;
    
    my $input=$from->stream();
    return unless defined $input;
    my $bytes_read=0;
	while($bytes_read < $bytes){
	    my $new_bytes=read($input,$content,$bytes-$bytes_read,$bytes_read);
	    $bytes_read+=$new_bytes;
	
	    last unless defined $new_bytes;
	}
    print "number of bytes = $bytes_read \n" if $main::debugging;
    
    $self->content($content);

    return $self;
}    

###### response->title()
###
###	Return the title of an HTML page, if it has one.
###	Returns the URL if the content-type is not HTML.
###
sub title {
    my($self)=@_;
    return unless $self->is_response();
    return $self->{'_title'} if defined $self->{'_title'};

    my $ttl  = $self->request->url();
    my $type = $self->content_type();
    return unless $type;
    return $ttl unless $type =~ m:text/html:;

    $self->read_content();	# === hopefully this is idempotent!
				# === really ought to just read header.

    my $page = $self->content();

    if ($page =~ m:<title>(.*)</title>:ig) { $ttl = $1; }
    $self->{'_title'} = $ttl;
    return $ttl;
}

#utilityfunction to turn post content into hash

# unescape URL-encoded data
sub unescape {
    my($todecode) = @_;
    $todecode =~ tr/+/ /;	# pluses become spaces
    $todecode =~ s/%([0-9a-fA-F]{2})/pack("c",hex($1))/ge;
    return $todecode;
}


sub compute_form_parameters{
    my($self)=@_;
    my $tosplit=$self->content();

    my(@pairs) = split('&',$tosplit);
    my($param,$value);
    my %hash;
    foreach (@pairs) {
	($param,$value) = split('=');
	$param = &unescape($param);
	$value = &unescape($value);
	$hash{$param}=$value; #careful losing multiple values
	print "  $param=$value.\n"  if $main::debugging;
    }
    $$self{parameters}=\%hash;
    
}
sub parameters{
    my ($self,$parameters)=@_;
    $$self{parameters}=$parameters if defined $parameters;
    return $$self{parameters};
    
}

1;
